### ISSUE TITLE
XPathEvaluator on CRM App for Tablet and Mobile

### ISSUE THREAD

**Issue by carfup** [con-ssc/XrmServicetoolkit#2]
Hello,

Your library is awesome, nothing to report so far until now :).
We are currently migrating to the CRM Online environment, and of course will use the CRM app for phones / tablets which go with it.

I've figured out that the app is working perfectly on phones / tablets running on iOS or Android but on Windows / Windows phone, when i load a form, i have "thousands" popup saying that XPathEvaluator is undefined. as you can see here : http://i.imgur.com/4zxVjPe.png

Do you have any solution ?

Thanks a lot,
Clément

**Comment by con-ssc** [comment]
Hello,
First of, it is not my lib actually, I just pushed the maintainer a bit to make it available here so contribution is easier.
What browser do you have and what version of the lib? Also, with what function do you have the problem? You could also send me the form script if you like so I can have a look what might cause this.

Regards,

Stefan
________________________________

**Comment by carfup** [comment]
Hello Stefan,

Thanks for your quick reply, indeed github is usefull to exchange about your solution.

To answer your questions, as i said in my first message, it happens only in the Phone / Tablet app on windows and windows phone, i assume it's using Edge browser by default ?
The version of your library i'm using is the 2.2, i also tried to update it to 2.2.1 but same result.

I'm not using the XPathEvaluator directly in my own code, the only reference of it i have in the javascripts loaded with the form is the following function in your lib: 
`var selectSingleNode = function (node, xpathExpr) {
        if (typeof (node.selectSingleNode) != "undefined") {
            return node.selectSingleNode(xpathExpr);
        }
        else {
            var xpe = new XPathEvaluator();
            var results = xpe.evaluate(xpathExpr, node, nsResolver, XPathResult.FIRST_ORDERED_NODE_TYPE, null);
            return results.singleNodeValue;
        }
    };`

That's why i thought about your library.
I just tried to be curious what was happening if i comment out this function, and of course i get a "selectSingleNode is undefined". again, i'm not calling this method from my custom code so ...

Thanks a lot for your support,
Clément

**Comment by con-ssc** [comment]
It seems to me, that this a copied over from the SDK. XPathEvaluator should be available if it is Edge. Although, the mobile Version may differ.
Please try the following workaround
```JavaScript
function selectSingleNode(node, xpathExpr) {
    if (typeof (node.selectSingleNode) != "undefined") {
        return node.selectSingleNode(xpathExpr);
    } else if(XPathEvaluator) {
        var xpe = new XPathEvaluator();
        var results = xpe.evaluate(xpathExpr, node, nsResolver, XPathResult.FIRST_ORDERED_NODE_TYPE, null);
        return results.singleNodeValue;
    } else {
        var doc = (node.nodeType != 9 ? node.ownerDocument : node);
        if (typeof doc.evaluate != "undefined") {
            var result = doc.evaluate(xpathExpr, node, nsResolver, XPathResult.FIRST_ORDERED_NODE_TYPE, null);
            return (result !== null ? result.singleNodeValue : null);
        }
    }
};
```

**Comment by carfup** [comment]
No more luck with this new code.
And yes you are right, this code comes from the SDK. That's pretty weird that it doesn't work by "default"...

**Comment by con-ssc** [comment]
Do you get the same error?

**Comment by carfup** [comment]
No in fact, i forgot to mention that, the error is slightly different.

I added this to your code :

else if (typeof(XPathEvaluator) != "undefined") {

and the new error is :
Error Code:-2147220989
which is about an invalid argument but i don't know how to or if it's
possible to debug directly from the app since i can't reproduce it on any
browsers ...

2016-05-03 15:01 GMT+02:00 con-ssc <notifications@github.com>:

**Comment by con-ssc** [comment]
okay, on what platform do you have this error?

**Comment by carfup** [comment]
The CRM App https://www.microsoft.com/en-us/store/apps/microsoft-dynamics-crm/9wzdncrfjbcm on a Windows 10 laptop and Windows Phone 10 cellphone.
all other OS are working

**Comment by con-ssc** [comment]
Do you have a Visual Studio that you could connect to the phone or the labtop? Otherwise maybe use the Field Medic app.

**Comment by con-ssc** [comment]
Also, maybe check if jquery is loaded on the form. If it is below v2 than it might lead to errors also.

**Comment by carfup** [comment]
Yeah i've found out that VS 2015 has more capabilities to directly debug the App than 2012 i was working on .... 
Yes the jquery loaded is the latest version.

I've found things with a colleague earlier.
It turned out that the error was poping as soon as the function selectSingleNode return a null value.
I tricked it to return something like "<entity name='role'><entity>" instead.

It's "fixing" the selectSingleNode error but have something different now :), at least i have some progress.

I will keep investigating and get back to you if i have a solution.

Thanks!
